lib/presentation/widget/shared_space_details/add_shared_space_member/add_shared_space_member_viewmodel.dart

```diff
@@ -119,19 +123,59 @@ class AddSharedSpaceMemberViewModel extends BaseViewModel {
           .execute(sharedSpaceId,
               AddSharedSpaceMemberRequest(accountId, sharedSpaceId, role.sharedSpaceRoleId))
-          .then((result) => store.dispatch(SharedSpaceAction(result)));
-      store.dispatch(SharedSpaceDetailsGetAllSharedSpaceMembersAction(
-          await _getAllSharedSpaceMembersInteractor.execute(sharedSpaceId)));
+          .then((result) => result.fold(
+                  (failure) => store.dispatch(AddSharedSpaceMembersAction(
+                      Left<Failure, Success>(AddSharedSpaceMemberFailure(AddMemberFailed())))),
+                  (success) {
+                store.dispatch(AddSharedSpaceMembersAction(Right(success)));
+              }));
+
+      store.dispatch(_getSharedSpaceMembersAction(sharedSpaceId));
     });
   }
 
-  OnlineThunkAction _getSharedSpaceRolesAction() {
+  OnlineThunkAction _updateSharedSpaceMemberRoleAction(
+      SharedSpaceId sharedSpaceId, AccountId accountId, SharedSpaceRoleName newRole) {
     return OnlineThunkAction((Store<AppState> store) async {
-      final roles = (await _getAllSharedSpaceRolesInteractor.execute())
-          .map((result) => result is SharedSpaceRolesViewState ? result.roles : null)
-          .getOrElse(() => []);
+      final rolesList = store.state.sharedSpaceState.rolesList;
+      if (rolesList.isEmpty) {
+        store.dispatch(UpdateSharedSpaceMembersAction(
+            Left<Failure, Success>(UpdateSharedSpaceMemberFailure(RolesListNotFound()))));
+        return;
+      }
+
+      final role = rolesList.firstWhere((_role) => _role.name == newRole, orElse: () => null);
+      if (role == null) {
+        store.dispatch(UpdateSharedSpaceMembersAction(
+            Left<Failure, Success>(UpdateSharedSpaceMemberFailure(SelectedRoleNotFound()))));
+        return;
+      }
+
+      await _updateSharedSpaceMemberInteractor
+          .execute(sharedSpaceId,
+              UpdateSharedSpaceMemberRequest(accountId, sharedSpaceId, role.sharedSpaceRoleId))
+          .then((result) => result.fold(
+                  (failure) => store.dispatch(UpdateSharedSpaceMembersAction(
+                      Left<Failure, Success>(UpdateSharedSpaceMemberFailure(UpdateRoleFailed())))),
+                  (success) {
+                store.dispatch(UpdateSharedSpaceMembersAction(Right(success)));
+                store.dispatch(_getSharedSpaceMembersAction(sharedSpaceId));
+              }));
+    });
+  }
 
-      store.dispatch(SharedSpaceGetSharedSpaceRolesListAction(roles));
+  OnlineThunkAction _getSharedSpaceMembersAction(SharedSpaceId sharedSpaceId) {
+    return OnlineThunkAction((Store<AppState> store) async {
+      store.dispatch(SharedSpaceDetailsGetAllSharedSpaceMembersAction(
+          await _getAllSharedSpaceMembersInteractor.execute(sharedSpaceId)));
     });
   }
+
+  @override
+  void onDisposed() {
+    store.dispatch(CleanAddSharedSpaceMembersStateAction());
+    store.dispatch(CleanUpdateSharedSpaceMembersStateAction());
```

Comment:
you want to reset the state of `addMember` and `listMember` screen?

Reply:
yes